a8n: Implement `previewCampaignPlan`

mrnugget opened the pull request:
This is part of https://github.com/sourcegraph/sourcegraph/issues/6085 and implements the GraphQL `previewCampaign` mutation.

It validates arguments, creates a `CampaignPlan` and multiple `CampaignJob`s (one for each repo yielded by the `CampaignType`s search query).

The single `CampaignType` that's currently implemented is `"comby"`, which does search&replace across repositories. 

What's in this PR already works: a plan is created, jobs are launched, executed and their `Diff` field is updated. 

There are a few things missing that I want to address in follow-up PRs in order to make reviewing easier and to unblock @felixfbecker and @eseliger: 

- The diff produced by reading json-lines from the `replacer` service is not a valid multi-file diff. While it can be parsed again by `diff.ParseMultiFile` the produced `[]*diff.FileDiff` only has a single element. That's enough to use the API for now, which is why I don't want it to block the merging of this PR. I will address this in a separate PR that also adds tests for `comby` (which right now only does a single HTTP request which we likely need to change)
- `CampaignPlan`s are not cleaned up yet. I already discussed this with @tsenart.
- The `CampaignPlan.status` is still a dummy implementation.

mrnugget commented on enterprise/pkg/a8n/run/comby.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (c *comby) searchQuery() string { return c.args.ScopeQuery }
+func (c *comby) generateDiff(repo api.RepoName, commit api.CommitID) (string, error) {
```
Add a `context.Context`

mrnugget commented on enterprise/pkg/a8n/run/run.go:
```diff
@@ -0,0 +1,169 @@
+			}
+
+			job.Diff = diff
+		}(context.Background(), r.ct, job)
```
```suggestion
		}(ctx, r.ct, job)
```

tsenart commented on enterprise/pkg/a8n/resolvers/resolver.go:
```diff
@@ -406,101 +402,40 @@ func (r *Resolver) PreviewCampaignPlan(ctx context.Context, args graphqlbackend.
-			CampaignPlanID: plan.ID,
-			StartedAt:      time.Now().UTC(),
-		}
+	if err := r.store.CreateCampaignPlan(ctx, plan); err != nil {
```
In the context of #6043, while working on validating that a campaign can't be created with a campaign plan that hasn't yet finished, it became clear that we need to do one of two things here:

1. Ensure that creating a campaign plan and its jobs is atomic (i.e. wrapped in a transaction).
2. Add a `finished_at` column to the `campaign_plans` table which would get set when all associated `campaign_jobs` are finished.

With this code as it is we have no way to know at campaign creation time if a campaign plan is finished, since there may be jobs that haven't been created yet.

I will proceed with the rest of the ticket and punt implementing this validation until we decide on the solution to this.

tsenart replied:
See here my first attempt at doing this validation at the database layer, where I realised the above problem: https://github.com/sourcegraph/sourcegraph/blob/768364335f0fd4a483e9923e50187f9abe4eee69/migrations/1528395611_validate_campaign_plan_completion_with_trigger.up.sql#L10

mrnugget replied:
Good point, thanks for bringing this up! My mental model assumes this:

> Ensure that creating a campaign plan and its jobs is atomic (i.e. wrapped in a transaction).

I also think that this invariant — "once a campaign plan has been created, all of its job have too" — is also the simplest one to hold for now.

mrnugget replied:
Implemented this as a transaction.

mrnugget commented:
I added the last commit 41f8d0e after our [discussion in Slack](https://sourcegraph.slack.com/archives/CMMTWQQ49/p1573048400043600) about skipping repositories without a default branch.

rvantonder commented on enterprise/pkg/a8n/run/campaign_type.go:
```diff
@@ -0,0 +1,142 @@
+	type rawCodemodResult struct {
+		URI  string `json:"uri"`
+		Diff string
+	}
```
You can now use https://sourcegraph.com/github.com/sourcegraph/sourcegraph@88b17fb69767800d3cadd4e3016d2ef5ce4b6b6e/-/blob/internal/comby/types.go#L61-65 instead of defining this inline.

mrnugget replied:
Thanks!

mrnugget commented:
Just added 374fc893fcf0a46b7290e2cd97fecf504e923787 to temporarily fix the problem of multi-file diffs.

The problem is that the `go-diff` package doesn't parse multi-file diffs correctly if they're missing a `diff ...` separator line between the single file diffs. `git apply` and `patch`, though, can handle these perfectly fine. So that is a "bug" in `go-diff` that I want to fix.

Until that's fixed and to make working with this branch easier, this temporary fix injects a `diff ...` line between the file diffs. 

The resulting multi-file diff can then be parsed by `go-diff` in the GraphQL backend and return multiple file diffs per repository. That gives us the correct response. Here's an example from a campaign that ran across multiple repositories and replaced text in zero or more files in each repository:

<details>
```json
{
  "data": {
    "previewCampaignPlan": {
      "__typename": "CampaignPlan",
      "id": "Q2FtcGFpZ25QbGFuOjEx",
      "type": "comby",
      "arguments": "{\"scopeQuery\":\"repo:gorilla\",\"matchTemplate\":\"example.com\",\"rewriteTemplate\":\"sourcegraph.com\"}",
      "status": {
        "completedCount": 0,
        "pendingCount": 99,
        "state": "ERRORED",
        "errors": [
          "this is just a skeleton api"
        ]
      },
      "changesets": {
        "nodes": [
          {
            "diff": {
              "fileDiffs": {
                "totalCount": 0,
                "nodes": []
              }
            },
            "repository": {
              "name": "ghe.sgdev.org/sourcegraph/gorilla-sessions",
              "url": "/ghe.sgdev.org/sourcegraph/gorilla-sessions"
            }
          },
          {
            "diff": {
              "fileDiffs": {
                "totalCount": 4,
                "nodes": [
                  {
                    "oldPath": "example_cors_method_middleware_test.go",
                    "newPath": "example_cors_method_middleware_test.go",
                    "hunks": [
                      {
                        "body": " \tfmt.Println(rw.Header().Get(\"Access-Control-Allow-Origin\"))\n \t// Output:\n \t// GET,PUT,PATCH,OPTIONS\n-\t// http://example.com\n+\t// http://sourcegraph.com\n }\n",
                        "oldRange": {
                          "startLine": 33
                        },
                        "newRange": {
                          "startLine": 33
                        }
                      }
                    ]
                  },
                  {
                    "oldPath": "README.md",
                    "newPath": "README.md",
                    "hunks": [
                      {
                        "body": " \n Setting the same matching conditions again and again can be boring, so we have a way to group several routes that share the same requirements. We call it \"subrouting\".\n \n-For example, let's say we have several URLs that should only match when the host is `www.example.com`. Create a route for that host and get a \"subrouter\" from it:\n+For example, let's say we have several URLs that should only match when the host is `www.sourcegraph.com`. Create a route for that host and get a \"subrouter\" from it:\n \n ```go\n r := mux.NewRouter()\n",
                        "oldRange": {
                          "startLine": 150
                        },
                        "newRange": {
                          "startLine": 150
                        }
                      },
                      {
                        "body": " s.HandleFunc(\"/articles/{category}/{id:[0-9]+}\", ArticleHandler)\n ```\n \n-The three URL paths we registered above will only be tested if the domain is `www.example.com`, because the subrouter is tested first. This is not only convenient, but also optimizes request matching. You can create subrouters combining any attribute matchers accepted by a route.\n+The three URL paths we registered above will only be tested if the domain is `www.sourcegraph.com`, because the subrouter is tested first. This is not only convenient, but also optimizes request matching. You can create subrouters combining any attribute matchers accepted by a route.\n \n Subrouters can be used to create domain or path \"namespaces\": you define subrouters in a central place and then parts of the app can register its paths relatively to a given subrouter.\n \n",
                        "oldRange": {
                          "startLine": 165
                        },
                        "newRange": {
                          "startLine": 165
                        }
                      }
                    ]
                  },
                  {
                    "oldPath": "route.go",
                    "newPath": "route.go",
                    "hunks": [
                      {
                        "body": " // For example:\n //\n //     r := mux.NewRouter()\n-//     r.Host(\"www.example.com\")\n+//     r.Host(\"www.sourcegraph.com\")\n //     r.Host(\"{subdomain}.domain.com\")\n //     r.Host(\"{subdomain:[a-z]+}.domain.com\")\n //\n",
                        "oldRange": {
                          "startLine": 284
                        },
                        "newRange": {
                          "startLine": 284
                        }
                      },
                      {
                        "body": " // It will test the inner routes only if the parent route matched. For example:\n //\n //     r := mux.NewRouter()\n-//     s := r.Host(\"www.example.com\").Subrouter()\n+//     s := r.Host(\"www.sourcegraph.com\").Subrouter()\n //     s.HandleFunc(\"/products/\", ProductsHandler)\n //     s.HandleFunc(\"/products/{key}\", ProductHandler)\n //     s.HandleFunc(\"/articles/{category}/{id:[0-9]+}\"), ArticleHandler)\n",
                        "oldRange": {
                          "startLine": 455
                        },
                        "newRange": {
                          "startLine": 455
                        }
                      }
                    ]
                  },
                  {
                    "oldPath": "mux_test.go",
                    "newPath": "mux_test.go",
                    "hunks": [
                      {
                        "body": " \t\t\tpath:            \"\",\n \t\t\tquery:           \"foo=bar&baz=ding\",\n \t\t\tpathTemplate:    `/api`,\n-\t\t\thostTemplate:    `www.example.com`,\n+\t\t\thostTemplate:    `www.sourcegraph.com`,\n \t\t\tqueriesTemplate: \"foo=bar,baz=ding\",\n \t\t\tqueriesRegexp:   \"^foo=bar$,^baz=ding$\",\n \t\t\tshouldMatch:     true,\n",
                        "oldRange": {
                          "startLine": 782
                        },
                        "newRange": {
                          "startLine": 782
                        }
                      },
                      {
                        "body": " \t\t\tpath:            \"\",\n \t\t\tquery:           \"foo=bar&baz=ding\",\n \t\t\tpathTemplate:    `/api`,\n-\t\t\thostTemplate:    `www.example.com`,\n+\t\t\thostTemplate:    `www.sourcegraph.com`,\n \t\t\tqueriesTemplate: \"foo=bar,baz=ding\",\n \t\t\tqueriesRegexp:   \"^foo=bar$,^baz=ding$\",\n \t\t\tshouldMatch:     true,\n",
                        "oldRange": {
                          "startLine": 796
                        },
                        "newRange": {
                          "startLine": 796
                        }
                      }
                    ]
                  }
                ]
              }
            },
            "repository": {
              "name": "ghe.sgdev.org/sourcegraph/gorilla-mux",
              "url": "/ghe.sgdev.org/sourcegraph/gorilla-mux"
            }
          },
          {
            "diff": {
              "fileDiffs": {
                "totalCount": 1,
                "nodes": [
                  {
                    "oldPath": "x_net_proxy.go",
                    "newPath": "x_net_proxy.go",
                    "hunks": [
                      {
                        "body": " \n // AddFromString parses a string that contains comma-separated values\n // specifying hosts that should use the bypass proxy. Each value is either an\n-// IP address, a CIDR range, a zone (*.example.com) or a host name\n+// IP address, a CIDR range, a zone (*.sourcegraph.com) or a host name\n // (localhost). A best effort is made to parse the string and errors are\n // ignored.\n func (p *proxy_PerHost) AddFromString(s string) {\n",
                        "oldRange": {
                          "startLine": 94
                        },
                        "newRange": {
                          "startLine": 94
                        }
                      }
                    ]
                  }
                ]
              }
            },
            "repository": {
              "name": "ghe.sgdev.org/sourcegraph/gorilla-websocket",
              "url": "/ghe.sgdev.org/sourcegraph/gorilla-websocket"
            }
          },
          {
            "diff": {
              "fileDiffs": {
                "totalCount": 0,
                "nodes": []
              }
            },
            "repository": {
              "name": "ghe.sgdev.org/sourcegraph/gorillalabs-sparkling",
              "url": "/ghe.sgdev.org/sourcegraph/gorillalabs-sparkling"
            }
          }
        ]
      }
    }
  }
}
```
</details>

tsenart commented on enterprise/pkg/a8n/run/runner.go:
```diff
@@ -0,0 +1,227 @@
+			RepoID:         repoID,
+			Rev:            rev,
+		}
+		if err := tx.CreateCampaignJob(ctx, job); err != nil {
```
I think we need a unique constraint on `(campaign_plan_id, repo_id, rev)`

mrnugget replied:
Will follow up in separate PR

tsenart commented on enterprise/pkg/a8n/run/runner.go:
```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return nil, err
+	}
+
```
```suggestion
      jobs := make([]*a8n.CampaignJob, 0, len(rs))
```

tsenart commented on enterprise/pkg/a8n/run/runner_test.go:
```diff
@@ -0,0 +1,408 @@
+
+			wantJobs := tc.wantJobs(plan, rs, revs)
+			jobIgnore := cmpopts.IgnoreFields(a8n.CampaignJob{}, "ID")
+			if !cmp.Equal(haveJobs, wantJobs, jobIgnore) {
```
```suggestion
			if diff := cmp.Diff(haveJobs, wantJobs, jobIgnore); diff != "" {
```

mrnugget commented on enterprise/pkg/a8n/resolvers/resolver.go:
```diff
@@ -402,105 +399,43 @@ func (r *Resolver) PreviewCampaignPlan(ctx context.Context, args graphqlbackend.
-			CampaignPlanID: plan.ID,
-			StartedAt:      time.Now().UTC(),
-		}
+	runner := run.New(r.store, campaignType, graphqlbackend.SearchRepos, nil)
```
We discussed this on a call: I'll play around a bit with the runner interface in the next few days.
